traffic_ops/traffic_ops_golang/trafficstats/trafficstats.go

```diff
@@ -0,0 +1,691 @@
+		api.HandleErr(w, r, tx, errCode, userErr, sysErr)
+		return
+	}
+
+	client, err := inf.CreateInfluxClient()
+	if err != nil {
+		if err == sql.ErrNoRows {
+			errCode = http.StatusServiceUnavailable
+			userErr = errors.New("No InfluxDB servers available!")
+			sysErr = userErr
+		} else {
+			errCode = http.StatusInternalServerError
+			userErr = nil
+			sysErr = err
+		}
+		api.HandleErr(w, r, tx, errCode, userErr, sysErr)
+		return
+	} else if client == nil {
+		userErr = errors.New("Traffic Stats is not configured!")
+		sysErr = userErr
+		errCode = http.StatusServiceUnavailable
+		api.HandleErr(w, r, tx, errCode, userErr, sysErr)
+		return
+	}
+	defer (*client).Close()
+
+	var dsTenant uint
+	if dsTenant, err = dsTenantIDFromXMLID(c.DeliveryService, tx); err != nil {
+		sysErr = err
+
+		if err == sql.ErrNoRows {
+			userErr = fmt.Errorf("No such Delivery Service: %s", c.DeliveryService)
+			errCode = http.StatusNotFound
+		} else {
+			errCode = http.StatusInternalServerError
+		}
+
+		api.HandleErr(w, r, tx, errCode, userErr, sysErr)
+		return
+	}
+
+	authorized, err := tenant.IsResourceAuthorizedToUserTx(int(dsTenant), inf.User, tx)
+	if err != nil {
+		api.HandleErr(w, r, tx, http.StatusInternalServerError, nil, err)
+		return
+	} else if !authorized {
+		// If the Tenant is not authorized to use the resource, then we DON'T tell them that.
+		// Instead, we don't disclose that such a Delivery Service exists at all - in keeping with
+		// the behavior of /deliveryservices
+		// This is different from what Perl used to do, but then again Perl didn't check tenancy at
+		// all.
+		userErr = fmt.Errorf("No such Delivery Service: %s", c.DeliveryService)
+		sysErr = fmt.Errorf("GetDSStats: unauthorized Tenant (#%d) access", inf.User.TenantID)
+		errCode = http.StatusNotFound
+		api.HandleErr(w, r, tx, errCode, userErr, sysErr)
+		return
+	}
+
+	var errBuilder strings.Builder
+	resp := APIResponse{
+		tc.TrafficStatsResponse{
+			Source:  tc.TRAFFIC_STATS_SOURCE,
+			Version: tc.TRAFFIC_STATS_VERSION,
+			Series:  nil,
+			Summary: nil,
+		},
+	}
+
+	// TODO: as above, this could be done on TO itself, thus sending only one synchronous request
+	// per hit on this endpoint, rather than the current two. Not sure if that's worth it for large
+	// data sets, though.
+	if !c.ExcludeSummary {
+		summary, messages, err := getSummary(client, &c, inf.Config.ConfigInflux.DSDBName)
+
+		errBuilder := strings.Builder{}
+		if messages != nil && len(messages) > 0 {
+			errBuilder.Write([]byte("Messages: ["))
+			for _, m := range messages {
+				errBuilder.WriteString(m.Level)
+				errBuilder.WriteRune(':')
+				errBuilder.WriteString(m.Text)
+				errBuilder.Write([]byte(", "))
+			}
+			errBuilder.WriteRune(']')
+		}
+
+		if err != nil {
+			sysErr = fmt.Errorf("Getting summary response from Influx: %v - %s", err, errBuilder.String())
+			api.HandleErr(w, r, tx, http.StatusBadGateway, nil, sysErr)
```

Comment:
I think this should be 500

Reply:
Again I think a 502 is more appropriate here:

> _"The 502 (Bad Gateway) status code indicates that the server, while acting as a gateway or proxy, received an invalid response from an inbound server it accessed while attempting to fulfill the request."_

That perfectly describes this situation

Reply:
is it ? do we return 502 on mysql errors? The user is not really asking you to proxy the request anywhere or understand if you are or not. Per the API contract the user is just asking for stats on delivery services and expects a certain return structure.

I have not worked with or on an API when it returns a 502 when it gets a bad response or can not connect to a backend db

Reply:
The Traffic Ops database is a part of traffic ops, so failure to communicate with it is considered a failure of the Traffic Ops server. But Traffic Stats is not just an implementation detail of Traffic Ops, it's a documented Traffic Control component in its own right, with its own API. The user of the API does, in fact, know that the data is retrieved this way because the response body says so, and because that fact is documented.

Reply:
I still think 502 is very very weird here and disagree with it. But Ill defer to the majority of the team though what we should do. end of the day something broke and its a 5xx error

Reply:
I think any server error in TO needs to be a 500 with no additional information, for security reasons. We don't currently have Self Service and random people on the internet logging in and making DSes, but as we move to Self Service, this will be a necessity. Better to do it now, than have to go back and change it later.

When there's a server-side error (not caused by client input, i.e. a 4xx), attackers can use that information to better craft an attack. For example, suppose Influx is dying (alas, not uncommon). If an attacker sees a 502 from this endpoint, they can guess that, and DOS just this endpoint, to overwhelm Influx even worse. Or, they can DOS this endpoint when it's working correctly until they get a 502, and discover exactly how many requests/second are required to DOS Influx.

The important thing is that the right person can find the information to fix the problem. Which they can: it's in the log. For a user error, we give the user as much information as possible about what they did wrong. For a server error, we give the user as little information as possible in case they're an attacker, and log as much as possible, so a good user can call Ops who can then find and fix the issue. If an Ops person encounters the error, they immediately look in the log.

Reply:
_Any_ server error needs to be a 500? Or just things that I'd otherwise want to be 502s? Either way, there'll need to be an issue created for making the rest of TO compliant with that.

Reply:
Any. And yes, TO used to be compliant, but I'm sure people have added things since.
And not just 500, but specifically not sending any info. I'm sure we also have endpoints returning info about server errors to users, which also need fixed.

Reply:
#3946